---
title: "RNA seq Report"
author: "Niel Infante"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: slidy_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=F)


# Error logging
log <- file(snakemake@log[[1]], open="wt")
sink(log)
sink(log, type="message")

# Load needed packages
library(tidyverse)
library(DESeq2)
library("RColorBrewer")
library(ggrepel)
library(gplots)
library(pheatmap)
library(knitr)
library(DT)
library(kableExtra)
library(ggiraph)

# Read dds
exp <- snakemake@params$exp
dds <- readRDS(paste0("results/", exp, "/deseq/dds.rds"))

vsd <- vst(dds, blind=F)


# For testing only
#vsd <- varianceStabilizingTransformation(dds, blind=F)


# Select genes which are ost informative,
# ie those that have the biggest difference in expression
asvd <- assay(vsd)
info <- rowMax(asvd) - rowMin(asvd)
minInfo <- info[order(info, decreasing = T)][500]
vsd <- vsd[info > minInfo,]


# Read Metadata and source config
meta_file_name <- snakemake@config$metadata_file
meta <- read_tsv(meta_file_name)

#snakemake@source(paste0('../', exp, '/deseq/', 'config.R'))
source(paste0('results/', exp, '/deseq/config.R'))

# Read results
res <- read_tsv(paste0('results/', exp, '/deseq/results.txt'))
sig <- read_tsv(paste0('results/', exp, '/deseq/significant.txt'))

# Read CV
cv <- read_tsv(paste0('results/', exp, '/deseq/basemean_cutoff.txt'))
cut_value <- cv$CV[1]
max_y <- cv$MY[1]
```

## Table of Contents
<UL>
	<LI><a href="#(3)">Differential Expression Results</a>
	<LI><a href="#(8)">Visualizations of Differential Expression</a>
	<LI><a href=#(14)>Gene Ontology</a>
	<LI><a href=#(21)>KEGG Pathways</a>
</UL>

## Results
This is the format of the results table, only 500 out of `r dim(res)[1]` genes are shown. The full list, with more columns, is available in the file results.txt.
```{r results}

tableToShow <- res %>% select(GeneName, baseMean, log2FoldChange, padj, Biotype, Cluster)

tableToShow <- tableToShow[1:500,]

tableToShow$padj <- signif(tableToShow$padj, digits=4)


#kable(res[1:5,])

datatable(tableToShow, filter='bottom', options = list(pageLength=10, lengthMenu = c(5, 10, 25))) %>% 
	formatRound('log2FoldChange', 3) %>% 
	formatRound('baseMean', 0)


```


## Explanation of Columns

```{r explain_col}

columns <- c('GeneID',
						 'GeneName',
						 'baseMean',
						 'log2FoldChange',
						 'lfcSE',
						 'stat',
						 'pvalue',
						 'padj',
						 'biotype',
						 'Sample Columns',
						 'Sample TPM',
						 'meanTPM',
						 'Cluster')

meanings <- c('Ensembl Gene ID',
							'The name of the gene',
							'Average expression of the gene across all samples',
							'log2 of the fold change of the gene',
							'Standard Error of the fold change',
							'The test statistic of the gene, used to calculate p value. Useful in some downstream analyses',
							'Unadjusted p value',
							'p value adjusted using Benjaminni-Hockberg correction',
							'The Biotype of the gene. Note that currently only one biotype is picked for each gene. Some genes may have more than one biotype',
							'Normalized counts for each sample',
							'Transcripts Pre Million for each sample',
							'Average of TPM across samples',
							'Expression cluster for filtering')

exp_col <- data.frame(Column=columns, Meaning=meanings)

kable(exp_col) %>% 
	kable_styling('striped', full_width = F) %>%
  column_spec(1, bold = F, border_right = T) %>%
  column_spec(2, width = "50em")	

```
See <a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html">here</a> for a discussion of different count normalization methods, and when to choose which.

## Fold Change

```{r sanity}

title=paste(res[1,]$GeneName, "\nFold Change:",res[1,]$log2FoldChange)
plotCounts(dds, gene=res[1,]$GeneID, intgroup = PCA_Group, main=res[1,]$GeneName, 
					 sub=paste('FC:', format(res[1,]$log2FoldChange, digits=2)), pch=19)


```

Fold change is reported as log2 (Fold Change). This graph gives you an idea of what a positive versus negative fold change means.

## Expression Cutoff

```{r expr_cut}

#ggplot(res, aes(x=log2(meanTPM + 1), fill=Cluster)) + 
#	geom_density() + scale_fill_manual(values=c('#00BA38', '#F8766D'))

ggplot(res, aes(x=log2(meanTPM+1))) + geom_density() + 
	geom_vline(xintercept=cut_value, color='red') +
	geom_text(aes(x=cut_value, y=0.7 * max_y, label='   Expression Cutoff',    hjust='left'), color='red')



```

Genes are kept or excluded based on expression clustering.

```{r alt_exp_cut, eval=F}

# An alternate plot for the expression cut off

ggplot(res, aes(x=log2(meanTPM + 1), y=..count.. + 1) ) + 
	ggridges::geom_density_line( aes(fill=Cluster, color=Cluster), alpha=0.25 ) + 
	ggridges::geom_density_line( data=select(res, -Cluster) , alpha=0.25) +
	scale_color_manual(values=c('#00BA38', '#F8766D')) + 
  scale_fill_manual(values=c('#00BA38', '#F8766D')) 



```


## Significant Genes
A total of `r dim(sig)[1]` genes were counted as significant in this test. Genes are counted as significant if the adjusted p value is below 0.05, there is a fold change of at least 1.5, and they are in the high expression cluster.

The top significant genes are:

```{r significant}


tableToShow <- sig %>% select(-GeneID, -stat, -lfcSE, -pvalue)

tableToShow$padj <- signif(tableToShow$padj, digits=4)

#kable(res[1:5,])

datatable(tableToShow, filter='bottom', options = list(pageLength=10, lengthMenu = c(5, 10, 25), scrollX=T)) %>% 
	formatRound('log2FoldChange', 3) %>% 
	formatRound('baseMean', 0)


```



<!-- This is the easy way to do a MA plot. Below
     is the hard way
## MA Plot
```{r maplot}
plotMA(dds)
```
-->

## MA Plot
```{r maplot_interactive, out.width="60%"}

sigUp   <- sig %>% filter(log2FoldChange > 0)
sigDown <- sig %>% filter(log2FoldChange < 0)
sigUp   <- sigUp %>% mutate(onClick = sprintf("window.open(\"%s%s\")","https://www.ensembl.org/id/", GeneID))
sigDown <- sigDown %>% mutate(onClick = sprintf("window.open(\"%s%s\")","https://www.ensembl.org/id/", GeneID))

#sigUp   <- sigUp %>% mutate(onClick = sprintf("window.open(\"%s%s\")","https://www.genecards.org/cgi-bin/carddisp.pl?gene=", GeneName))
#sigDown <- sigDown %>% mutate(onClick = sprintf("window.open(\"%s%s\")","https://www.genecards.org/cgi-bin/carddisp.pl?gene=", GeneName))
# Would need ENTREZID, but this might be best
#sigDown <- sigDown %>% mutate(onClick = sprintf("window.open(\"%s%s\")","https://www.ncbi.nlm.nih.gov/gene/", ENTREZID))


	
# Make the plot
ma_plot <- ggplot(res, aes(log2(baseMean), log2FoldChange)) + geom_point(color='grey', alpha=0.5, pch=1)  +
	geom_point_interactive(data=sigUp, color='darkmagenta', aes(tooltip=GeneName, onclick=onClick)) +
	geom_point_interactive(data=sigDown, color='darkgoldenrod2', aes(tooltip=GeneName, onclick=onClick)) +
	theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
	labs(x='Log Expression', y='Log Fold Change')

# Make it interactive
girafe(code = print(ma_plot) )

```



<!--  Remove this section 
## Sample Distance

```{r sample_heat}

select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing=TRUE)[1:100]
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)

distsRL <- dist(t(assay(vsd)))
mat <- as.matrix(distsRL)
rownames(mat) <- colnames(mat) <- with(colData(dds), paste(Graph_Display))

heatmap.2(mat, trace="none", col = rev(hmcol), margin=c(8, 8), density.info = 'none')

```

This is a heatmap of sample distances. Dark indicates the samples are closer together, based on the expression of the 100 most expressed genes.

-->

## Cluster

```{r cluster}
plot(hclust(dist(t(assay(vsd)))), label=with(colData(dds), paste(Graph_Display)), main='Dendrogram', xlab='', sub='')

```


## PCA
```{r pca}


print(plotPCA(vsd, intgroup=c(PCA_Group)))
```

A Principle Component plot of the data. Each point is a sample.


## PCA with Names
```{r pca_names}

p <- plotPCA(vsd, intgroup=c(PCA_Group))
p <- p + geom_text_repel(aes(x=PC1, y=PC2, 
				label=with(colData(dds), paste(Graph_Display))), 
				point.padding = unit(2,"points"))
print(p)

```

The same PCA plot, with added names



##  Volcano

```{r volcano}

par(pch = 16)
with(res, plot(log2FoldChange, -log10(pvalue), main = "Volcano plot"))
with(subset(res, padj < 0.05), points(log2FoldChange, -log10(pvalue), col = "red"))
with(subset(res, abs(log2FoldChange) > 2), points(log2FoldChange, -log10(pvalue),  col = "orange"))

with(subset(res, padj < 0.05 & abs(log2FoldChange) > 2), points(log2FoldChange,  -log10(pvalue), col = "green"))

# Add legend
legend("topleft", legend = c("FDR<0.05", "|LFC|>2", "both"), pch = 16, col = c("red", "orange", "green"))

```


## Expression Heatmap

```{r pheatmap, message=F, warning=F}

rld <- rlog(dds, blind = F)

# Get 15 up, 15 down 

up <- head(order(res$log2FoldChange),15)
up <- res[up,'GeneID']


dn <- head(order(res$log2FoldChange, decreasing = T),15)
dn <- res[dn,'GeneID']

#get <- c(up,dn)
get <- c(unlist(up),unlist(dn))


rld15 <- rld[get]
#rld15 <- rld[rld@rowRanges@partitioning@NAMES %in% get, ]

rn <- data.frame(GeneID=rownames(rld15))
rn <- left_join(rn, res)

rownames(rld15) <- rn$GeneName

mat <- assay(rld15)
mat <- mat - rowMeans(mat)     # Get difference from mean for each gene
df <- as.data.frame(colData(rld15)[,PCA_Group])
names(df) <- PCA_Group

# Need the same names in both the df and the matrix
rownames(df) <- colData(rld15)[,c('Graph_Display')]
colnames(mat) <- colData(rld15)[,c('Graph_Display')]


pheatmap(mat, cutree_rows = 2, cutree_cols = 2, annotation_col = df, angle_col = "45")

```

Heatmap shows the 15 most upregulated, and 15 most downregulated genes. Color indicates how much the expression of the gene in the samples is different from the mean expression of that gene.

<!--- Include all the GO analysis -->
```{r GO, child='../../scripts/GO_report.rmd'}

```

<!--- Include all the KEGG analysis -->
```{r KEGG, child='../../scripts/KEGG_report.rmd'}

```


